### PR TITLE
CRYP-40: implement email notifications

### PR DESCRIPTION
**Pull request by jason-gerard**
JIRA: https://cryptify.atlassian.net/browse/CRYP-40

## Changes
- Add notification abstract template method
- Add email notification implementation
- Add function call and integration to btc-edge, eth-edge will be done in another PR

### CODE REVIEW

File: packages/btc-edge/src/services/transaction_watcher_service.ts [hunk 1]
```diff
@@ -43,6 +46,9 @@ export class TransactionWatcherService {
                 const txAddress = (res as WSTransaction).x.hash;
                 const transactions = await this.soChainGateway.getTransactionsByTxAddress(txAddress);
                 await this.transactionsRepository.save(transactions);
+
+                // Asynchronously send the notification to the users
+                this.notificationService.sendTransactionNotifications(transactions, getCurrencyType(txAddress));
```
Comment (Andre-Ibrahim) [inline packages/btc-edge/src/services/transaction_watcher_service.ts:51]: Why is this only in Btc-edge?

also just leaving a note that this needs to be only if user allows notification.

**Review by Andre-Ibrahim — APPROVED** [review]
from what I see there doesn't seem to be any setup for the mailer other than `yarn install` correct?

Everything else looks good just left a comment as a note for a future user story!

File: packages/common/src/utils/notifications/abstract_notification_service_template_method.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                        // template method design pattern. This allows us to centralize all the pre-processing and
+                        // formatting in a single place while allowing for a module notification system. This can be
+                        // easily expanded to support sms and push notifications by simply extending this abstract
+                        // class and implementing only the specific functionality to send out a single notification
+                        // for that platform
+                        await this.sendNotification({
+                            to: wallet.user.email,
+                            title,
+                            body,
+                        });
```
Comment (alexbara2000) [inline packages/common/src/utils/notifications/abstract_notification_service_template_method.ts:45]: This code smells

**Review by alexbara2000 — APPROVED** [review]
Looks very good! The only thing, would be the nesting in the notification service but not a big deal.

File: packages/common/src/utils/notifications/abstract_notification_service_template_method.ts (hunk 2 above)
Comment (jason-gerard) [reply packages/common/src/utils/notifications/abstract_notification_service_template_method.ts:45]: Yea the nesting could be improved upon thank you for your comment. I'll be updating it later on.

File: packages/btc-edge/src/services/transaction_watcher_service.ts (hunk 1 above)
Comment (jason-gerard) [reply packages/btc-edge/src/services/transaction_watcher_service.ts:51]: Yep just adding a POC here, then in another pr ill add it in eth-edge and with the db updates etc

**Pull request merged**